### PR TITLE
Bulk array validation

### PR DESCRIPTION
**Pull request by tlively**
Add validation rules (prose and formalism) for `array.fill`, `array.copy`,
`array.init_data`, and `array.init_elem`. As part of the rules for `array.copy`,
extend the `matches` relation to storage types.

### CODE REVIEW

**Comment by tlively** [comment]
Current dependencies on/for this PR:
* main
  * **PR #401** <a href="https://app.graphite.dev/github/pr/WebAssembly/gc/401" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/WebAssembly/gc/401?utm_source=stack-comment).

**Review by rossberg — APPROVED** [review]
LGTM modulo comments.

File: document/core/valid/matching.rst
Comment (rossberg) [inline document/core/valid/matching.rst:152]: Did you want to delete those?

File: document/core/valid/instructions.rst
Comment (rossberg) [inline document/core/valid/instructions.rst:696]: This seems to be missing a closing tick.

File: document/core/valid/matching.rst
Comment (tlively) [reply document/core/valid/matching.rst:152]: nope, thanks

File: document/core/valid/matching.rst
Comment (tlively) [reply document/core/valid/matching.rst]: I've now also removed the macros that will be added back in by #382.

**Comment by tlively** [comment]
We'll wait on #382 to land this, since it no longer builds independently.

**Pull request merged**
